src/include/optimizer/abstract_optimizer.h

```diff
@@ -43,6 +44,8 @@ class AbstractOptimizer {
       const std::unique_ptr<parser::SQLStatementList> &parse_tree) = 0;
 
   virtual void Reset(){};
+
+  concurrency::Transaction *txn = nullptr;
```

Comment:
And I also think txn is optimizer's quality. You should know what txn you optimizer each time. Why I cannot declare it in high level?